README.md
# react-md-spinner

![Screenshot](./docs/images/repo-banner.gif)

[![Build Status](https://dev.azure.com/wadackel/react-md-spinner/_apis/build/status/tsuyoshiwada.react-md-spinner?branchName=master)](https://dev.azure.com/wadackel/react-md-spinner/_build/latest?definitionId=1&branchName=master)
[![npm version](https://img.shields.io/npm/v/react-md-spinner.svg)](https://www.npmjs.com/package/react-md-spinner)

> Material Design spinner components for React.js.

Live example: https://tsuyoshiwada.github.io/react-md-spinner/

## Table of Contents

- [Installation](#installation)
- [Features](#features)
- [Getting Started](#getting-started)
  - [Basic Usage](#basic-usage)
  - [Server-Side Rendering](#server-side-rendering)
    - [Example](#example)
- [Props](#props)
  - [`size`](#size)
  - [`borderSize`](#bordersize)
  - [`duration`](#duration)
  - [`color1`](#color1)
  - [`color2`](#color2)
  - [`color3`](#color3)
  - [`color4`](#color4)
  - [`singleColor`](#singlecolor)
- [API](#api)
  - [`ssrBehavior`](#ssrbehavior)
    - [As string output](#as-string-output)
    - [As React Components](#as-react-components)
- [ChangeLog](#changelog)
- [Contributing](#contributing)
- [Available Scripts](#available-scripts)
  - [`yarn test`](#yarn-test)
  - [`yarn lint`](#yarn-lint)
  - [`yarn format`](#yarn-format)
  - [`yarn build`](#yarn-build)
  - [`yarn storybook`](#yarn-storybook)
- [License](#license)

## Installation

You can install the [react-md-spinner](https://www.npmjs.com/package/react-md-spinner) from [npm](https://www.npmjs.com/).

```bash
$ npm i -S react-md-spinner
# or
$ yarn add react-md-spinner
```

## Features

- :rocket: You can start using with zero configuration!
- :wrench: Support to change of color, size, border and animation speed.
- :sparkling_heart: It can also be used in single color.
- :globe_with_meridians: Support Server-Side Rendering.

## Getting Started

### Basic Usage

Because it is made of 100% inline styles, you can start using it right away without setting.

```typescript
import React from "react";
import MDSpinner from "react-md-spinner";

export const SpinnerExample: React.FC = () => (
  <div>
    <MDSpinner />
  </div>
);
```

### Server-Side Rendering

The following is an example of Server-Side Rendering.  
Please checkout [examples](./examples/) directory for details.

The point is to use `ssrBehavior`.

#### Example

**Note:** The following is pseudo code.

**Client-Side:**

```typescript
import React from "react";
import { render } from "react-dom";
import App from "./App";

render(<App />, document.getElementById("app"));
```

**Server-Side:**

```typescript
import { ssrBehavior } from "react-md-spinner";

// ...

const html = (root: JSX.Element) => `<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    ${ssrBehavior.getStylesheetString()}
  </head>
  <body>
    <div id="app">${renderToString(root)}</div>
    <script defer src="/client.js"></script>
  </body>
</html>`;

app.get("/", (_req, res) => {
  res.status(200).send(`<!doctype html>${renderer(<App />)}`);
});
```

**App:**

```typescript
import React from "react";
import MDSpinner from "react-md-spinner";

export const App: React.FC = () => (
  <div>
    <MDSpinner />
  </div>
);
```

## Props

You can use the following Props. All Props are Optional!

### `size`

**type:** `number`  
**default:** `28`

Set the size (diameter) of the spinner circle.

### `borderSize`

**type:** `number`  
**default:** `undefined`

Set the spinner border size of. By default, the appropriate size is calculated according to the value of `size`.

### `duration`

**type:** `number`  
**default:** `1333`

Set the animation duration (ms) of the spinner.

### `color1`

**type:** `string`  
**default:** !`rgb(66, 165, 245)`

The color of the spinner. Can be set to any valid CSS string (hex, rgb, rgba).

### `color2`

**type:** `string`  
**default:** `rgb(239, 83, 80)`

Same as above.

### `color3`

**type:** `string`  
**default:** `rgb(253, 216, 53)`

Same as above.

### `color4`

**type:** `string`  
**default:** `rgb(76, 175, 80)`

Same as above.

### `singleColor`

**type:** `string`  
**default:** `undefined`

Same as above. Use this if the spinner should be in only one single color. The settings (props) for `color1` ~ `4` will be ignored by setting this `singleColor` property.

## API

### `ssrBehavior`

In Server-Side Rendering you need to inject `@keyframes` inside the `<head>`.  
`react-md-spinner` provides utilities to handle them.

- `ssrBehavior.getStylesheetString(): string`
- `ssrBehavior.getStylesheetComponent(): React.ReactNode`

#### As string output

```typescript
import { ssrBehavior } from "react-md-spinner";

const html = () => `<!doctype html>
  <head>
    ${ssrBehavior.getStylesheetString()}
  </head>
  <body>
    <div id="app">
      // React stuff here
    </div>
  </body>
</html>`;
```

#### As React Components

```typescript
import React from "react";
import { ssrBehavior } from "react-md-spinner";

const Html: React.FC = () => (
  <html>
    <head>{ssrBehavior.getStylesheetComponent()}</head>
    <body>
      <div id="app">{/* React stuff here */}</div>
    </body>
  </html>
);
```

## ChangeLog

See [CHANGELOG.md](./CHANGELOG.md)

## Contributing

We are always welcoming your contribution :clap:

1. Fork (https://github.com/tsuyoshiwada/react-md-spinner) :tada:
1. Create a feature branch :coffee:
1. Run test suite with the `$ yarn test` command and confirm that it passes :zap:
1. Commit your changes :memo:
1. Rebase your local changes against the `master` branch :bulb:
1. Create new Pull Request :love_letter:

## Available Scripts

### `yarn test`

Run unit test using Jest.

### `yarn lint`

Run Lint of source code using ESLint.

### `yarn format`

Run formatting using Prettier and ESLint's Fixer.

### `yarn build`

Run build of TypeScript code.

### `yarn storybook`

Run Storybook.

## License

[MIT © tsuyoshiwada](./LICENSE)


CHANGELOG.md
# CHANGELOG

## 1.0.0 (2019-06-09)

### New Features

- Rewrite with TypeScript!
- Small bundle size!!
- Build process using rollup and babel. (support `module` field)
- Storybook examples.

### Breaking Changes

- Remove `useAgent` props. (always vendor prefixes)
- Remove `useAgent` argument on `ssrBehavior` API.

Changed to use `stylis` instead of `inline-style-prefixer`. Along with that, we now always have a vendor prefix for styles in `@keyframes`.

These changes have the benefit of reducing the bundle size.

### Internal Changes

- TSLint -> ESLint.
- Using Prettier.
- Using Dependabot.

## 0.4.0 (2019-06-01)

- Replace deprecated lifecycle hook. Thanks [@ChrisBrownie55](https://github.com/ChrisBrownie55) !

## 0.3.0 (2018-07-08)

- Add `borderSize` props.  
  Thanks [@dbalas](https://github.com/dbalas) !

## 0.2.5 (2017-07-22)

- Changes unused Component types from `undefined` to `{}`  
  Thanks [@nbgraham](https://github.com/nbgraham) !

## 0.2.4 (2017-05-22)

- Add type declaration file for typescript [#9](https://github.com/tsuyoshiwada/react-md-spinner/pull/9)  
  Thanks [@devholic](https://github.com/devholic) !!

## 0.2.3 (2017-05-08)

- [rofrischmann/inline-style-prefixer](https://github.com/rofrischmann/inline-style-prefixer) update to `3.0.3`. [#8](https://github.com/tsuyoshiwada/react-md-spinner/pull/8)  
  Thanks [@MichaelDeBoey](https://github.com/MichaelDeBoey).
- Bump 'react-highlight' dependency.
- Add dependencies [caniuse-api](https://github.com/nyalab/caniuse-api) package.

## 0.2.2 (2017-04-21)

- Use 'prop-types' package instead of React.PropTypes. ([#6](https://github.com/tsuyoshiwada/react-md-spinner/pull/6))  
  Thanks [@MichaelDeBoey](https://github.com/MichaelDeBoey).

## 0.2.1 (2017-03-09)

- Update dev/dependencies.
  - [rofrischmann/inline-style-prefixer](https://github.com/rofrischmann/inline-style-prefixer) update to `3.0.1`.

## 0.2.0 (2017-02-02)

### New features

- Full support for Server-Side Rendering. [#3](https://github.com/tsuyoshiwada/react-md-spinner/issues/3).
  - Thanks [@cescoferraro](https://github.com/cescoferraro)!
- Add `ssrBehavior` utilities.
  - Please checkout [examples](./examples/) directory for details.

### Minor changes

- Add [examples](./examples/).
- Update docs.
- Add this ChangeLog.
- Change to `docs` from `demo` for gh-pages.
- Update dependencies.

## 0.1.0 (2016-09-23)

### New features

- Add `userAgent` props. (for Server-Side Rendering)
  - Thanks [@ryan-codingintrigue](https://github.com/ryan-codingintrigue)!

### Minor changes

- Update dependencies.

## 0.0.3 (2016-07-19)

- Fix specification of the numeric string on size prop.
- Update dependencies.

## 0.0.2 (2016-07-03)

- Fix npm files in package.json.

## 0.0.1 (2016-07-03)

- First release.


demo/entry.js
import React, { Component, PropTypes } from "react";
import ReactDOM from "react-dom";
import Highlight from "react-highlight";
import injectTapEventPlugin from "react-tap-event-plugin";
import MuiThemeProvider from "material-ui/styles/MuiThemeProvider";
import RaisedButton from "material-ui/RaisedButton";
import MDSpinner from "../";


injectTapEventPlugin();

function Rect({ color }) {
  return <div className="rect" style={{ backgroundColor: color }} />;
}

Rect.propTypes = {
  color: PropTypes.string.isRequired
};

class App extends Component {
  render() {
    /* eslint-disable max-len */
    return (
      <div>
        {/* hero */}
        <div className="hero">
          <div className="hero__container container">
            <div className="hero__text">
              <h1 className="hero__logo"><img src="./images/logo@2x.png" width="400" /></h1>
              <p className="hero__lead">
                Material Design spinner components for React.js.<br />
                Because it is made of 100% inline styles, you can start using it right away!
              </p>
              <div className="hero__install">
                <Highlight className="bash">
                  {"$ npm install react-md-spinner"}
                </Highlight>
              </div>
              <ul className="hero__button-group">
                <li className="hero__button">
                  <RaisedButton
                    primary
                    label="Document on GitHub"
                    href="https://github.com/tsuyoshiwada/react-md-spinner"
                  />
                </li>
              </ul>
            </div>
            <div className="hero__spinner">
              <MDSpinner size={60} />
            </div>
          </div>
        </div>

        {/* contents */}
        <div className="contents">
          <div className="container">
            {/* size */}
            <div className="block">
              <h2 className="block__title">Size</h2>
              <div className="block__row">
                <div className="block__col">
                  <div className="block__body">
                    <p>You can easily specify a custom size(default: 28px). Therefore, it is possible to use in a flexible scene.</p>
                  </div>
                </div>
                <div className="block__col">
                  <Highlight className="html">
                    {`<div>
  <MDSpinner size={16} />
  <MDSpinner size={30} />
  <MDSpinner size={50} />
  <MDSpinner size={70} />
  <MDSpinner size={100} />
</div>`}
                  </Highlight>
                </div>
              </div>
              <div className="block__result">
                <MDSpinner size={16} />
                <MDSpinner size={30} />
                <MDSpinner size={50} />
                <MDSpinner size={70} />
                <MDSpinner size={100} />
              </div>
            </div>

            {/* Duration */}
            <div className="block">
              <h2 className="block__title">Duration</h2>
              <div className="block__row">
                <div className="block__col">
                  <div className="block__body">
                    <p>You can change the Duration of the animation (default: 1333ms).<br />In the following example, we dropped the speed of the icon.</p>
                  </div>
                </div>
                <div className="block__col">
                  <Highlight className="html">
                    {`<div>
  <MDSpinner duration={2000} />
</div>`}
                  </Highlight>
                </div>
              </div>
              <div className="block__result">
                <MDSpinner duration={2000} />
              </div>
            </div>

            {/* Colors */}
            <div className="block">
              <h2 className="block__title">Colors</h2>
              <div className="block__row">
                <div className="block__col">
                  <div className="block__body">
                    <p>You can change the base color of 1 to 4.</p>
                    <p><strong>Default:</strong></p>
                    <ul>
                      {["rgb(66, 165, 245)", "rgb(239, 83, 80)", "rgb(253, 216, 53)", "rgb(76, 175, 80)"].map((color, index) =>
                        <li key={index}>color{index + 1}: <Rect color={color} /> {color}</li>
                      )}
                    </ul>
                  </div>
                </div>
                <div className="block__col">
                  <Highlight className="html">
                    {`<div>
  <MDSpinner
    color1="#e91e63"
    color2="#673ab7"
    color3="#009688"
    color4="#ff5722"
  />
</div>`}
                  </Highlight>
                </div>
              </div>
              <div className="block__result">
                <MDSpinner
                  color1="#e91e63"
                  color2="#673ab7"
                  color3="#009688"
                  color4="#ff5722"
                />
              </div>
            </div>

            {/* Single Color */}
            <div className="block">
              <h2 className="block__title">Single Color</h2>
              <div className="block__row">
                <div className="block__col">
                  <div className="block__body">
                    <p>Colorful spinner also nice. but it is also possible to change to a single color.</p>
                    <p>If you specify a singleColor, value of color1 ~ 4 are ignored.</p>
                  </div>
                </div>
                <div className="block__col">
                  <Highlight className="html">
                    {`<div>
  <MDSpinner singleColor="#03a9f4" />
</div>`}
                  </Highlight>
                </div>
              </div>
              <div className="block__result">
                <MDSpinner singleColor="#03a9f4" />
              </div>
            </div>
          </div>
        </div>

        {/* Footer */}
        <footer className="footer">
          <div className="footer__container container">
            <p className="footer__copyright">
              react-md-spinner.js licensed under MIT.<br />
              Made by <a href="https://github.com/tsuyoshiwada">tsuyoshiwada</a>.
              <MDSpinner
                size={14}
                singleColor="#009688"
                style={{
                  marginLeft: 10
                }}
              />
            </p>
          </div>
        </footer>
      </div>
    );
    /* eslint-enable max-len */
  }
}

ReactDOM.render(
  <MuiThemeProvider>
    <App />
  </MuiThemeProvider>,
  document.getElementById("app")
);
